---
title: "Scanning sequences for miRNA binding sites and exploring matches with scanMiR"
author: 
- name: Pierre-Luc Germain
  affiliation: D-HEST Institute for Neuroscience, ETH & Lab of Statistical Bioinformatics, UZH
- name: Michael Soutschek
  affiliation: Lab of Systems Neuroscience, D-HEST Institute for Neuroscience, ETH
- name: Fridolin Gross
  affiliation: Lab of Systems Neuroscience, D-HEST Institute for Neuroscience, ETH
package: scanMiR
output:
  BiocStyle::html_document
abstract: |
  This vignettes explores scanMiR's scanning functions.
vignette: |
  %\VignetteIndexEntry{1_scanning}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
library(BiocStyle)
library(Biostrings)
# only for testing
devtools::load_all("/mnt/schratt/miRNA_KD/fg_test/scanMiR")
# library(scanMiR)
```

# Scanning
## Background
`r Biocpkg("scanMiR")` can be used to identify potential binding sites given a set of miRNAs and a set of transcripts. Furthermore, it determines the type of binding site and, given a `KdModel` object, the predicted affinity of the site.

## Basic Scan
The main function used for determining matches of miRNA seeds in a given set of sequences is `findSeedMatches`. It accepts a set of DNA sequences either as a character vector or as a [DNAStringSet](https://bioconductor.org/packages/release/bioc/html/Biostrings.html). Seeds can be given either as a character vector or as a `KdModelList`. The following example illustrates the basic usage:

```{r}
# generate a set of mock sequences and seeds
seqs <- getRandomSeq(n=5)
seeds <- c("AAACCAC", "AAACCUU")

# run the scan
matches <- findSeedMatches(seqs, seeds, verbose = FALSE)
```

By default, a [GRanges](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) object is returned. Apart from the position of the matches, it provides information on the match type. Setting `ret = "data.frame"` returns the same information as a data.frame. The scan can be restricted to canonical sites using `onlyCanonical = TRUE`.

## Further Options

### ORF length
If the transcript sequences are provided as a `DNAStringSet`, one may specify the length of the open reading frame region of the transcripts in order to distinguish between matches in the ORF and 3'UTR regions.
```{r}
seqs <- DNAStringSet(seqs)
mcols(seqs)$ORF.length <- sample(500:800, length(seqs))
```

### Shadow and Overlapping Matches
The `shadow` argument can be used to take into account the observation that sites within the first ~15 nucleotides of the 3'UTR show poor efficiency ([Grimson et al. 2007](https://www.sciencedirect.com/science/article/pii/S1097276507004078)). `findSeedMatches` will treat matches within the first `shadow` positions of the UTR in the same way as matches in the ORF region. If no information on ORF lengths is provided, it will simply ignore the first `shadow` positions. The default setting is `shadow = 0L`.

The parameter `minDist` can be used to specify the minimum distance between matches of the
same miRNA (default 1). If there are multiple matches within `minDist`, only the highest affinity match will be considered.

### Site Affinities
If the seeds are given in the form of a `KdModelList`, `findSeedMatches` also returns the predicted affinity value for each match. The `log_kd` column contains log(Kd) values multiplied by 1000, where KD is the predicted dissociation constant of miRNA:mRNA binding at the respective site. In addition, a column `p3.score` is returned, which carries information on 3p alignment (see Section @ref(sec:aggregating).

```{r}
# load KdModel
data(SampleKdModel)
kd_seeds <- KdModelList(list(SampleKdModel))
kd_matches <- findSeedMatches(seqs, kd_seeds, verbose = FALSE)
```


### Aggregation on the fly
With `ret = "aggregated"` one obtains a data.frame that contains the predicted repression for each sequence-seed-pair aggregated over all matches along with information about the types and number of matches. Parameters for aggregation can be specified using `agg.params`. For further details, see Section \@ref(sec:aggregating).

### Parallel Processing
To deal with large amounts of sequences and/or seeds, `findSeedMatches` supports multithreading using the `r Biocpkg("BiocParallel")` package. This can be activated by passing `BP = MulticoreParam(ncores)`. In addition, the number of seeds that is processed simultaneously can be restricted using the parameter `n_seeds` in order to avoid memory issues.

# Aggregating Sites {#sec:aggregating}
## Background
`r Biocpkg("scanMiR")` implements aggregation of miRNA sites based on the biochemical model from [McGeary, Lin et al. (2019)](https://dx.doi.org/10.1126/science.aav1741). This model first predicts the occupancy of AGO-miRNA complexes at each potential binding site as a function of the measured or estimated dissociation constants (Kds). It then assumes an additive effect of the miRNA on the basal decay rate of the transcript that is proportional to this occupancy.

The key parameters of this model are:

* `ag`: the relative concentration of unbound AGO-miRNA complexes
* `b`: the factor that multiplies with the occupancy and is added to the basal decay rate (can be interpreted as the additional repression caused by a single bound AGO)
* `c`: the penalty factor for sites that are found within the ORF region

The resulting repression value corresponds to the expected (log) fold-change in the expression of the transcript when compared to a background of non-specific binding.

While `b` and `c` are considered global (i.e. the same for different miRNAs and transcripts and also across experimental contexts), `ag` is expected to be different for each miRNA-transcript pair in a given experimental condition. However, as shown by [McGeary, Lin et al. (2019)](https://dx.doi.org/10.1126/science.aav1741), the model performance is robust to changes in `ag` over several orders of magnitude. Aggregation for all miRNA-transcript pairs for a given data set is therefore usually based on a single `ag` value. `r Biocpkg("scanMiR")` provides reasonable default parameter values that have been determined by fitting and validating the model using several experimental data sets.

## Basic Aggregation
Given a `GRanges` or data.frame of matches as returned by `findSeedMatches`, aggregation can be performed by the function `aggregateMatches`:
```{r}
agg_matches <- aggregateMatches(kd_matches)
```
This returns a data.frame with predicted repression values for each miRNA-transcript pair. By using `keepSiteInfo = TRUE`, one additionally obtains information on the number and type of sites. As discussed in 

## Further Options
### 3p alignment

### coef_utr and coef_orf

### Parallel Processing
Like `findSeedMatches` also `aggregateMatches` supports multithreading by passing `BP = MulticoreParam(ncores)`.

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
